rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,223 @@
+
+### Automatic deployments
+
+We will enable automatic deployments for the apps listed [here](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/hieradata_aws/common.yaml#L145), but only when they meet the safety criteria of this RFC. To help with this, a rough audit of "compatibility" can be found in [Appendix A: Compatibility](rfc-128/compatibility.md).
+
+- We can immediately enable automatic deployments for apps marked with a ✅.
+- We will undertake work to address all the issues (⚠) for apps marked with a ❌.
+
+> **We will disable CD for Publishing API**. This was only enabled on a trial basis, and requires non-trivial work to bring it up to the new standard. We should avoid setting a false precedent by leaving CD enabled.
+
+The following steps must be taken as part of enabling automatic deployments:
+
+- Restrict merge access to "GOV.UK Production" [[example](https://github.com/alphagov/govuk-saas-config/pull/55)].
+
+- Tag [relevant Smoke tests](https://github.com/alphagov/smokey/blob/5dc36fc78c12838ec1c8070d6fbc535ae5433b89/docs/writing-tests.md#filtering-by-app) to run for the app [[example](https://github.com/alphagov/smokey/pull/708/commits/91a5ce94b7b96d09f5db87e63fbc3e400788de00)].
+
+- Add a note in the [Release](https://github.com/alphagov/release) app to warn about the change.
+
+- Add a PR template to warn about the change [[example](https://github.com/alphagov/smart-answers/blob/3509ba3e68c62a206874c96111628dd65adae689/.github/pull_request_template.md)].
+
+  - Backfill the PR template as a comment on existing PRs.
+
+- Enable automatic deployments in a Puppet PR [[example](https://github.com/alphagov/govuk-puppet/pull/10590)].
+
+  - Comment on how the app meets the criteria of this RFC.
+
+> **We can be flexible with coverage**. We can defer adding internal repo tests if the code has no impact on the running app e.g. rake tasks [[1](https://github.com/alphagov/email-alert-api/blob/59fc71a58317ef2998f2c0ef102020da3ca9df96/lib/tasks/support.rake)]. But we should not defer removing any unused code.
+
+### Slack notifications
+
+We need to ensure we are notified about a failure in the automation, so that the automation is no worse than the current, manual process. This has already been done for the `Deploy_App_Downstream` job [[1](https://github.com/alphagov/govuk-puppet/pull/10532)] [[2](https://github.com/alphagov/govuk-puppet/pull/10601)].
+
+So that we cover the entire deployment pipeline, we will also update [the `Deploy_App` job](https://github.com/alphagov/govuk-puppet/blob/f57335f6027487b5b2d2205a35ceb411776f09ce/modules/govuk_jenkins/templates/jobs/deploy_app.yaml.erb) to post a Slack message in the #govuk-developers Slack channel, whenever it fails due to an unexpected error.
+
+### Delete [publishing-e2e-tests](https://github.com/alphagov/publishing-e2e-tests)
```

Comment:
We plan to use Dockerfiles in app projects, but we'll need to modify them. It's OK with me if you would like to delete the Dockerfiles, we'll bring them back when we need them.

Reply:
Ok great - that sounds good @bilbof. It seems like it would be more prudent to delete them until they become used. We can always put together a quick gist or dev docs document with links to everything before deletion for reference so the work doesn't have to be repeated whilst avoiding the risks of keeping files/resources that aren't used for anything.

Reply:
Thanks @bilbof and @kevindew! I'd like to make this clear in the consequence, by editing to: "They, along with all of their associated infrastructure [[1](https://github.com/alphagov/email-alert-api/blob/0b87e62288ddb1653451f84e1f36e17ce4e8e9dc/Dockerfile)] [[2](https://github.com/alphagov/govuk-app-deployment/blob/58fcfab4262da0ba06a568dd5e9bbd76be2f4f62/lib/docker_tag_pusher.rb)], can be safely...".

Is that OK?

Reply:
Does anyone use these tests locally for testing anything, or is there more of a reliance on making branch deploys to integration to check everything works (i.e. smoke tests pass) because of the hassle of getting these E2E tests running locally?

Reply:
Yes, definitely the latter: the E2E tests take _absolutely ages_ to spin up locally. I've never heard of anyone spinning them up locally, except to debug a problem with the tests themselves.

Although it's not equivalent, GOV.UK Docker does provide some support for manual testing of apps working together, with a much lower barrier to entry.

Reply:
Yes, removing all infrastructure sounds good.

Reply:
OK. I've added a commit for this.

@karlbaker02 I'm going to assume I've answered your question and resolve this. Feel free to comment back if you think there's still more to discuss.